File: Soe.Angular/TypeScript/Common/Models/ExtraFieldDTO.ts
import { ICompTermDTO, IExtraFieldDTO, IExtraFieldGridDTO, IExtraFieldRecordDTO, IExtraFieldValueDTO } from "../../Scripts/TypeLite.Net4";
import { SettingDataType, SoeEntityState, SoeEntityType, TermGroup_ExtraFieldType, TermGroup_ExtraFieldValueType } from "../../Util/CommonEnumerations";

export class ExtraFieldDTO implements IExtraFieldDTO {
    entity: SoeEntityType;
    extraFieldId: number;
    extraFieldRecords: IExtraFieldRecordDTO[];
    extraFieldValues: IExtraFieldValueDTO[];
    text: string;
    translations: ICompTermDTO[];
    type: TermGroup_ExtraFieldType;
    connectedEntity: number;
    connectedRecordId: number;
    externalCodesString: string;
}

export class ExtraFieldValueDTO implements IExtraFieldValueDTO {
    created: Date;
    createdBy: string;
    extraFieldId: number;
    extraFieldValueId: number;
    modified: Date;
    modifiedBy: string;
    sort: number;
    state: SoeEntityState;
    type: TermGroup_ExtraFieldValueType;
    value: string;
}

export class ExtraFieldGridDTO implements IExtraFieldGridDTO {
    accountDimId: number;
    accountDimName: string;
    extraFieldId: number;
    extraFieldValues: IExtraFieldValueDTO[];
    hasRecords: boolean;
    text: string;
    type: number;

    // Extensions
    typeName: string;

    get isText(): boolean {
        return this.type === TermGroup_ExtraFieldType.FreeText;
    }
    get isInteger(): boolean {
        return this.type === TermGroup_ExtraFieldType.Integer;
    }
    get isDecimal(): boolean {
        return this.type === TermGroup_ExtraFieldType.Decimal;
    }
    get isYesNo(): boolean {
        return this.type === TermGroup_ExtraFieldType.YesNo;
    }
    get isCheckbox(): boolean {
        return this.type === TermGroup_ExtraFieldType.Checkbox;
    }
    get isDate(): boolean {
        return this.type === TermGroup_ExtraFieldType.Date;
    }
    get isSingleChoice(): boolean {
        return this.type === TermGroup_ExtraFieldType.SingleChoice;
    }
    get isMultiChoice(): boolean {
        return this.type === TermGroup_ExtraFieldType.MultiChoice;
    }
}

export class ExtraFieldRecordDTO implements IExtraFieldRecordDTO {
    boolData: boolean;
    comment: string;
    dataTypeId: number;
    dateData: Date;
    decimalData: number;
    extraFieldId: number;
    extraFieldRecordId: number;
    extraFieldText: string;
    extraFieldType: number;
    extraFieldValues: IExtraFieldValueDTO[];
    _intData: number;
    recordId: number;
    strData: string;
    _value: string;

    toPlainObject(): object {
        return {
            ...this,
            intData: this.intData,
            value: this.value
        };
    }

    get value(): string {
        let value = "";

        switch (this.dataTypeId) {
            case SettingDataType.Boolean:
                if (this.boolData !== undefined && this.boolData !== null)
                    value = this.boolData.toString();
                break;
            case SettingDataType.Decimal:
                if (this.decimalData !== undefined && this.decimalData !== null)
                    value = this.decimalData.toFixed(4);
                break;
            case SettingDataType.Integer:
                if (this.intData !== undefined && this.intData !== null)
                    value = this.intData.toString();
                break;
            case SettingDataType.String:
                if (this.strData)
                    value = this.strData.toString();
                break;
            case SettingDataType.Date:
                if (this.dateData)
                    value = this.dateData.toLocaleDateString();
                break;
            default:
                break;
        }

        return value;
    }

    set value(value: string) {
        this._value = value;
    }

    get intData() {
        return this._intData;
    }
    set intData(value) {
        if (!value)
            value = 0;

        if (!Number.isInteger(value))
            this._intData = Number.parseInt(value.toString());
        else
            this._intData = value;
    }

    // Extensions
    isModified: boolean;
}

File: Soe.Common/DTO/ExtraFieldDTO.cs
using SoftOne.Soe.Common.Attributes;
using SoftOne.Soe.Common.Util;
using System;
using System.Collections.Generic;

namespace SoftOne.Soe.Common.DTO
{
    [TSInclude]
    public class ExtraFieldDTO
    {
        public ExtraFieldDTO()
        {
            Translations = new List<CompTermDTO>();
            ExtraFieldRecords = new List<ExtraFieldRecordDTO>();
            ExtraFieldValues = new List<ExtraFieldValueDTO>();
        }
        public int ExtraFieldId { get; set; }
        public int? SysExtraFieldId { get; set; }
        public SoeEntityType Entity { get; set; }
        public string Text { get; set; }
        public TermGroup_ExtraFieldType Type { get; set; }
        public int? ConnectedEntity { get; set; }
        public int? ConnectedRecordId { get; set; }
        public List<CompTermDTO> Translations { get; set; }
        public List<ExtraFieldRecordDTO> ExtraFieldRecords { get; set; }
        public List<string> ExternalCodes { get; set; } = new List<string>();
        public List<ExtraFieldValueDTO> ExtraFieldValues { get; set; }

        public string ExternalCodesString { get; set; }
        public int? CategoryGroupId { get; set; }
        public DateTime? Created { get; set; }
        public string CreatedBy { get; set; }
        public DateTime? Modified { get; set; }
        public string ModifiedBy { get; set; }
        public SoeEntityState State { get; set; }
    }

    [TSInclude]
    public class ExtraFieldValueDTO
    {
        public int ExtraFieldValueId { get; set; }
        public int ExtraFieldId { get; set; }

        public TermGroup_ExtraFieldValueType Type { get; set; }
        public string Value { get; set; }
        public int Sort { get; set; }
        public DateTime? Created { get; set; }
        public string CreatedBy { get; set; }
        public DateTime? Modified { get; set; }
        public string ModifiedBy { get; set; }
        public SoeEntityState State { get; set; }
    }

    [TSInclude]
    public class ExtraFieldGridDTO
    {
        public int ExtraFieldId { get; set; }
        public string Text { get; set; }
        public int Type { get; set; }
        public int AccountDimId { get; set; }
        public string AccountDimName { get; set; }
        public bool HasRecords { get; set; }
        public int? SysExtraFieldId { get; set; }

        // Extensions
        public List<ExtraFieldValueDTO> ExtraFieldValues { get; set; }
    }

    [TSInclude]
    public class ExtraFieldRecordDTO
    {
        public int ExtraFieldRecordId { get; set; }
        public int ExtraFieldId { get; set; }
        public int DataTypeId { get; set; }
        public string StrData { get; set; }
        public int? IntData { get; set; }
        public bool? BoolData { get; set; }
        public DateTime? DateData { get; set; }
        public decimal? DecimalData { get; set; }
        public string Comment { get; set; }
        public int RecordId { get; set; }

        public string ExtraFieldText { get; set; }
        public int ExtraFieldType { get; set; }

        public List<ExtraFieldValueDTO> ExtraFieldValues { get; set; }

        public string Value
        {
            get
            {
                string value = "";

                switch (this.DataTypeId)
                {
                    case (int)SettingDataType.Boolean:
                        if (this.BoolData.HasValue)
                            value = this.BoolData.Value.ToString();
                        break;
                    case (int)SettingDataType.Decimal:
                        if (this.DecimalData.HasValue)
                            value = this.DecimalData.Value.ToString("0.0000");
                        break;
                    case (int)SettingDataType.Integer:
                        if (this.IntData.HasValue)
                            value = this.IntData.Value.ToString();
                        break;
                    case (int)SettingDataType.String:
                        if (!String.IsNullOrEmpty(this.StrData))
                            value = this.StrData.ToString();
                        break;
                    case (int)SettingDataType.Date:
                        if (this.DateData.HasValue)
                            value = this.DateData.ToShortDateString();
                        break;
                    default:
                        break;
                }

                return value;
            }
        }

    }
}
